File: backend/app/models/character.py
from datetime import datetime
from sqlalchemy import Column, String, DateTime, ForeignKey
from sqlalchemy.orm import relationship

from ..database import Base


class Character(Base):
    __tablename__ = "characters"

    id = Column(String(36), primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    description = Column(String, nullable=True)
    avatar = Column(String(500), nullable=True)
    personality = Column(String, nullable=True)
    background = Column(String, nullable=True)
    relationships = Column(String, nullable=True)
    novel_id = Column(String(36), ForeignKey("novels.id", ondelete="CASCADE"), nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    novel = relationship("Novel", back_populates="characters")

File: frontend/src/services/character.ts
import api from "./api";

export type Character = {
  id: string;
  name: string;
  description?: string | null;
  avatar?: string | null;
  personality?: string | null;
  background?: string | null;
  relationships?: string | null;
  novel_id: string;
};

export async function listCharacters(novelId: string): Promise<Character[]> {
  const { data } = await api.get<Character[]>(`/novels/${novelId}/characters/`);
  return data;
}

export async function createCharacter(
  novelId: string,
  payload: Pick<Character, "name"> & Partial<Omit<Character, "id" | "novel_id">>
): Promise<Character> {
  const { data } = await api.post<Character>(`/novels/${novelId}/characters/`, payload);
  return data;
}

export async function deleteCharacter(novelId: string, characterId: string): Promise<void> {
  await api.delete(`/novels/${novelId}/characters/${characterId}`);
}

